voting/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_sdk::collections::Map;
+use near_sdk::{env, near_bindgen, AccountId, Balance, EpochHeight};
+use uint::construct_uint;
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+type Timestamp = u64;
+
+construct_uint! {
+    /// 256-bit unsigned integer.
+    // TODO: Revert back to 4 once wasm/wasmer bug is fixed.
+    // See https://github.com/wasmerio/wasmer/issues/1429
+    pub struct U256(8);
+}
+
+/// Voting contract for unlocking transfers. Once the majority of the stake holders agree to
+/// unlock transfer, the time will be recorded and the voting ends.
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct VotingContract {
+    /// How much each validator votes
+    votes: Map<AccountId, Balance>,
+    /// Map of account to their current stake
+    account_stake: Map<AccountId, Balance>,
+    /// Total voted balance so far.
+    total_voted_stake: Balance,
+    /// When the voting ended. `None` means the poll is still open.
+    result: Option<Timestamp>,
+    /// Epoch height when the contract is touched last time.
+    last_epoch_height: EpochHeight,
+}
+
+impl Default for VotingContract {
+    fn default() -> Self {
+        env::panic(b"Voting contract should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl VotingContract {
+    #[init]
+    pub fn new() -> Self {
+        VotingContract {
+            votes: Map::new(b"a".to_vec()),
+            account_stake: Map::new(b"s".to_vec()),
+            total_voted_stake: 0,
+            result: None,
+            last_epoch_height: 0,
+        }
+    }
+
+    fn resolve_votes(&mut self) {
+        let cur_epoch_height = env::epoch_height();
+        if cur_epoch_height != self.last_epoch_height {
+            for account_id in self.account_stake.keys().into_iter().collect::<Vec<_>>() {
+                let old_account_stake = self.account_stake.remove(&account_id).unwrap();
+                let account_current_stake = env::validator_stake(&account_id);
+                let account_voted_stake = self.votes.remove(&account_id).unwrap();
+                if account_current_stake > 0 {
+                    let new_account_voted_stake = (U256::from(account_voted_stake)
+                        * U256::from(account_current_stake)
+                        / U256::from(old_account_stake))
+                    .as_u128();
+                    self.total_voted_stake =
+                        self.total_voted_stake + new_account_voted_stake - account_voted_stake;
+                    self.votes.insert(&account_id, &new_account_voted_stake);
+                    self.account_stake
+                        .insert(&account_id, &account_current_stake);
+                }
+            }
+            self.check_result();
+            self.last_epoch_height = cur_epoch_height;
+        }
+    }
+
+    fn check_result(&mut self) {
```

Comment:
this should be public, because with the same votes - it can pass later when total stake decreases.

Reply:
I made `ping` public.